File: python/02_loops.py
#!/usr/bin/env python3

# Loops!
# Why do something once when you can do it multiple times?

someNumbers = [1, 4, 3, 12, 147]

for n in someNumbers:
	nSquared = n*n
	print("%d squared is %d" % (n, nSquared))

# Loop is a term used to describe a repeated sequence of instructions
# that are performed until a specific condition is met.  Sounds daunting, but
# the above code is an example of a loop in that we
#
# - squared numbers (repeated instructions)
# - until we ran out of numbers to square (a specific condition is met)
#
# Some might say the above is an example of iterating, but I see
# iteration and looping as one in the same.

# Another typical loop, counting from one number to another:

print("Count from 0 up to 10")
for i in range(0, 10):
  print(i)

# Or we can count down
print("Count from 10 down to 0")
for i in range(10, 0, -1):
  print(i)

# We've seen two examples of loops that are iterators (processing items in a list)
# Now let's look at one that is based a logic condition
print("Count from 10 down to 1 with a while loop")
x = 10
while x > 0:
  print(x)
  x = x - 1

# These are very simple examples of loops but in general loops
# are designed to "do something multiple times until a condition is met"

File: javascript/02_loops.js
#!/usr/bin/env node

// Loops in Javascript are similar to loops in Python
// but the syntax is different.

var someNumbers = [1, 4, 3, 12, 147];

console.log("Iterating (looping) through a list");
for (var n in someNumbers) {
	var nSquared = n*n;
	console.log(`${n} squared is ${nSquared}`);
}

// Unlike Python, Javascript for loops are typed as
// for (expression) { statement1; statement2; ... statementN; }

// Javascript also does not have a built-in range() function like
// Python did, but that's okay we can use this.  This is a
// classic for loop that one would encounter in C (except for the
// var keyword)
console.log("Counting up!");
for (var i = 0; i <= 10; i++) {
	console.log(i);
}

// The expression is broken into 3 parts
// - the starting condition (i = 0)
// - the looping condition (as long as the loop condition is true)
// - the increment statement
// On the first iteration i is set to 0
// If the loop condition is true (0 is less than or equal to 10),
// execute what is in the block
// Increment i by 1 (++ is shorthand for "add one")

// As you can see, incrementing is in the eye of the beholder,
// as it can be a decrement if you want it to be...
console.log("Counting down with a for loop");
for (var i = 10; i >= 0; i--) {
	console.log(i);
}

// Like Python, Javascript also has a while loop
console.log("Counting down with a while loop");
var x = 10;
while (x > 0) {
	console.log(x);
	x--;
}

// At this point you can see that Javascript and Python are
// similar in their ability to store variables and have looping
// constructs.  This is not by accident, as all languages
// share this basic idea.

